runtime/parachains/src/builder.rs

```diff
@@ -537,46 +594,51 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
 
 	/// Build a scenario for testing or benchmarks.
 	///
-	/// - `backed_and_concluding_cores`: Map from core/para id/group index seed to number of
-	/// validity votes.
-	/// - `dispute_sessions`: Session index of for each dispute. Index of slice corresponds to core.
-	/// The length of this must equal total cores used. Seed index for disputes starts at
-	/// `backed_and_concluding_cores.len()`, so `dispute_sessions` needs to be left padded by
-	/// `backed_and_concluding_cores.len()` values which effectively get ignored.
-	/// TODO we should fix this.
-	pub(crate) fn build(
-		self,
-		backed_and_concluding_cores: BTreeMap<u32, u32>,
-		dispute_sessions: &[u32],
-		includes_code_upgrade: Option<u32>,
-	) -> Bench<T> {
+	/// Note that this API only allows building scenarios where the `backed_and_concluding_cores`
+	/// are mutually exclusive with the cores for disputes. So
+	/// `backed_and_concluding_cores.len() + dispute_sessions.len()` must be less than the max
+	/// number of cores.
+	///
+	/// Some methods for configuring the builder:
+	///
+	/// * `set_backed_and_concluding_cores`: Set a map from core/para id seed to number of validity
+	/// votes.
+	/// * `set_dispute_sessions`: Set the session index of for each dispute. Index of slice
+	/// corresponds to a core, which is offset by the number of entries for
+	/// `backed_and_concluding_cores`. I.E. if `backed_and_concluding_cores` cores has 3 entries,
+	/// the first index of `dispute_sessions` will correspond to core index 3. There must be an
+	/// entry for each core with a dispute statement set.
+	/// *`set_code_upgrade`: Set to `Some` to include a code upgrade for all backed candidates.
+	/// The value within `Some` will be the byte length of the code.
+	pub(crate) fn build(self) -> Bench<T> {
```

Comment:
wild idea, feel free to discard: return `Result<Bench<T>, Bench<T>>` and determine integrity of the bench setup, the user can then determine via `unwrap_err` and `unwrap` if that is desired or not. Not sure how feasabile this is and what kind of checks we can make with reasonable effort.

Reply:
I think this is a cool idea, my main concern though is that its there is so many corner cases to consider that we are bound to miss some and instead it would just be better for the user to be diligent

Reply:
Maybe issue-ify this and come back in the future.

Reply:
https://github.com/paritytech/polkadot/issues/4412